akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterShardingImpl.scala

```diff
@@ -159,7 +159,7 @@ import akka.util.Timeout
       stopMessage: Option[M],
       settings: ClusterShardingSettings,
       extractor: ShardingMessageExtractor[E, M],
-      allocationStrategy: Option[ShardAllocationStrategy]): ActorRef[E] = {
+      allocationStrategy: Option[() => ShardAllocationStrategy]): ActorRef[E] = {
```

Comment:
@patriknw I know we discussed not doing this but I found using the API without this very cumbersome as typically a sharded entity will have an init function that may be called many times and then many instances of the allocation strategy would be created e.g. https://github.com/akka/akka-samples/pull/155#discussion_r348641601

Alternatively we could create an extension for creating dynamic allocation strategy instances

Reply:
I see the point, but not totally convinced it's worth adding/changing the api.

Do we gain much by starting the `DDataStateActor` eagerly. I understand that the purpose is to init the subscription early so that it has some state when the first allocation/rebalance request comes in.

Maybe the `DDataStateActor` can do a local `Get` to the replicator for the first request? That shouldn't add much latency and it's anyway assumed that the ddata Replicator has been started early.

The DDataStateActor would then also be started lazily by the first allocation/rebalance request.

Reply:
I think that'll work.

The other one was to intitialize it once the node becomes oldest, but then the allocation strategy would also need to know about roles.

Another option would be to add an initialize method to the API with a default no-op

Reply:
I still find it an awkward API from a user perspective as they need to make sure that their allocation strategy is safe to be instantiated many times on every node of the cluster as it is common to get the ShardProxy from multiple places rather than creating it once and passing it around.

Reply:
ofc it is fine for stateless implementations but that doesn't cover all use cases which we cater for by having the API use futures

Reply:
I'm just questioning if it's worth adding all these overloads to the API (and also in typed api). I don't think many users write custom allocation strategies.

Reply:
Maybe this can be an alternative:

```scala
object DynamicShardAllocationStrategy {

  def apply(system: ActorSystem, typeName: String, timeout: Timeout): ShardAllocationStrategy =
    new LazyShardAllocationStrategy(() => new DynamicShardAllocationStrategy(system, typeName)(timeout))
}

class LazyShardAllocationStrategy(factory: () => ShardAllocationStrategy) extends ShardAllocationStrategy {

  private var delegate: OptionVal[ShardAllocationStrategy] = OptionVal.None

  def start(): Unit =
    delegate = OptionVal.Some(factory())

  override def allocateShard(
      requester: ActorRef,
      shardId: ShardId,
      currentShardAllocations: Map[ActorRef, immutable.IndexedSeq[ShardId]]): Future[ActorRef] =
    delegate match {
      case OptionVal.Some(d) => d.allocateShard(requester, shardId, currentShardAllocations)
      case OptionVal.None    => throw new IllegalStateException
    }

  override def rebalance(
      currentShardAllocations: Map[ActorRef, immutable.IndexedSeq[ShardId]],
      rebalanceInProgress: Set[ShardId]): Future[Set[ShardId]] =
    delegate match {
      case OptionVal.Some(d) => d.rebalance(currentShardAllocations, rebalanceInProgress)
      case OptionVal.None    => throw new IllegalStateException
    }
}
```

and inside sharding check if instanceof LazyShardAllocationStrategy and call `start` to activate it.

Reply:
That looks a good approach.

Could we do it without the impl if users extend a StartableShardAllocationStrategy trait that extends ShardALlocationStrategy with a start method and call it? Leaving any laziness up to the impl

Reply:
I tried this in https://github.com/akka/akka/pull/28211/commits/42f4bfa28c6c65655570f73a4a013ea559b39d13 WDYT?

Reply:
Good

Reply:
Nope, I just hadn't removed everything. I'll clean this up now.